pkg/controllers/job/job_controller_actions.go

```diff
@@ -653,14 +652,29 @@ func (cc *jobcontroller) createJobIOIfNotExist(job *batch.Job) (*batch.Job, erro
 					continue
 				}
 				job.Spec.Volumes[index].VolumeClaimName = vcName
-				needUpdate = true
 				break
 			}
 			// TODO: check VolumeClaim must be set if VolumeClaimName is empty
 			if volume.VolumeClaim != nil {
 				if err := cc.createPVC(job, vcName, volume.VolumeClaim); err != nil {
 					return job, err
 				}
+				newJob, err := cc.vcClient.BatchV1alpha1().Jobs(job.Namespace).Update(context.TODO(), job, metav1.UpdateOptions{})
+				if err != nil {
+					klog.Errorf("Failed to update Job %v/%v for volume claim name: %v ",
+						job.Namespace, job.Name, err)
+					return job, err
+				}
+				job.ResourceVersion = newJob.ResourceVersion
+
+				job.Status.ControlledResources["volume-pvc-"+vcName] = vcName
+				newJob, err = cc.vcClient.BatchV1alpha1().Jobs(job.Namespace).UpdateStatus(context.TODO(), job, metav1.UpdateOptions{})
```

Comment:
Why call the UpdateStatus interface again here? The outer code will also call UpdateStatus to update the job status later, I think `job.Status.ControlledResources["volume-pvc-"+vcName] = vcName` is enough

Reply:
If not updated in a timely here, the next for loop may exit, resulting in the loss of records here. So I update the status after creating the PVC.

Reply:
OK got it. However, I'm concerned that this might lead to too many requests being sent to the apiserver during batch creation. It seems that the job must be updated during the for loop. If the update fails within the for loop and the job is not updated, it could potentially cause the function to trigger PVC creation again the next time it's entered. I think the design here needs to be refactored, but for now, I think it's ok. We should redesign when we meet performance problem. @Monokaix cc

Reply:
The update is only triggered when the PVC is created, and in other cases, it will not trigger the new update operation added here.

Reply:
We reviewed it again, it might be caused by other errors rather than not updating the job in time. Here, the `job.Status.ControlledResources["volume-pvc-"+vcName] = vcName` refers to the job in the cache, which has already been updated to the job controller's cache previously:
https://github.com/volcano-sh/volcano/blob/8d1658ac7c4bc6a1ecbf968cb09c1cd749516508/pkg/controllers/job/job_controller_actions.go#L887

So it might still be an error caused by another location
Subsequent triggers of syncJob will also directly retrieve from the cache, so it will have PVC information:
https://github.com/volcano-sh/volcano/blob/8d1658ac7c4bc6a1ecbf968cb09c1cd749516508/pkg/controllers/job/job_controller.go#L368

Reply:
I got it...

Reply:
> The clone only copy the JobInfo, which is an inner structure of scheduler, the pointer of the volcano job is still copied:
> 
> https://github.com/volcano-sh/volcano/blob/8d1658ac7c4bc6a1ecbf968cb09c1cd749516508/pkg/controllers/apis/job_info.go#L41

OK, I got it.